### ISSUE TITLE
Remove recognition of KISAO terms for characteristics of algorithms

### ISSUE THREAD

**Issue by jonrkarr** [sys-bio/tellurium#500]
While its helpful that tellurium tries to recognize KISAO terms for similar algorithms, tellurium's implementation seems strange to me. Many of the terms that tellurium recognizes are terms for characteristics of algorithms rather than terms for algorithms. Algorithm characteristics is a distinct part of the KISAO ontology from the terms for algorithms. These extra terms for algorithm characteristics could cause tellurium to substantially distort the intended meaning of a simulation. For example, tellurium appears to suggest it can execute any random algorithm. While SSA is a random algorithm, there are many other random algorithms which are not SSA. In reality, tellurium can only execute one specific random algorithm. This is different from Python which could be used to implement any random algorithm.

**Comment by jonrkarr** [comment]
I created a Python package to do this more systematically, more comprehensively for all of KiSAO, and more consistently across tools. 

The Python package just implements (fairly simple) queries which could easily be replicated in other languages for other simulation tools. 

Note, the versioning of the Python package is the same as that of KiSAO itself.

Installation
```
pip install kisao
```

This is a sketch of what would go into tellurium:
```python
from kisao import Kisao
from kisao.data_model import AlgorithmSubstitutionPolicy
from kisao.utils import get_perferred_substitute_algorithm

# load ontology
kisao = Kisao()

# terms for the algorithms implemented by tellurium
implemented_algorithms_in_preferred_order = [
    kisao.get_term('KISAO_0000019'),
    kisao.get_term('KISAO_0000086'),
    kisao.get_term('KISAO_0000032'),
    kisao.get_term('KISAO_0000029'),
    kisao.get_term('KISAO_0000408'),
]

# get the term for the algorithm requested in a SED-ML document
requested_algorithm = kisao.get_term('KISAO_0000088')

# choose a substitution policy (recommended: SIMILAR_VARIABLES)
substitution_policy=AlgorithmSubstitutionPolicy.SAME_MATH
substitution_policy=AlgorithmSubstitutionPolicy.SIMILAR_VARIABLES

# get a similar algorithm that tellurium implements to the one requested in the SED-ML document
alternative_algorithm = get_preferred_substitute_algorithm(requested_algorithm, implemented_algorithms_in_preferred_order,
    substitution_policy=substitution_policy)

# get the short id for this algorithm
alternative_algorithm.id.partition('#')[2]
>> 'KISAO_0000019'
```

**Comment by luciansmith** [comment]
This looks super helpful--thanks!

Would it be possible to allow calling these functions by kisao integer instead of just the whole kisao string?  i.e.

    kisao.get_term(88)

?

**Comment by jonrkarr** [comment]
I'll add something like that.

I can definitely make changes. This is a first stab at trying to support the reasoning over the ontology that simulation tools need.

**Comment by luciansmith** [comment]
Right, and it's perfect to put this in a library, so then every tool developer doesn't need to implement it themselves (or, more likely, run incorrect heuristics like we're doing now ;-)  Thank you!

**Comment by jonrkarr** [comment]
I'll push this to PyPI a later today. `get_term` will work as you outlined. The method will normalize the id and then fetch the term from the ontology. There's also a couple of functions (`Kisao.get_term_id`, `Kisao.get_normalized_id`  to get ids of terms in the various dialects used):
- `IdDialect.kisao`: `"KISAO:0000019"`
- `IdDialect.sedml`: `"KISAO_0000019"`
- `IdDialect.integer`: `19`

**Comment by jonrkarr** [comment]
Here's the simplified version now available from PyPI

```python
from kisao.data_model import IdDialect, AlgorithmSubstitutionPolicy
from kisao.utils import get_preferred_substitute_algorithm_by_ids

# algorithm requested to execute e.g., from SED-ML document
alg_id = 88 

# what tellurium implements
implemented_alg_ids = [19, 86, 32, 29, 408] 

# allowed degree of substitution -- allow user to set this
substitution_policy = AlgorithmSubstitutionPolicy.SIMILAR_VARIABLES

#
exec_id = get_preferred_substitute_algorithm_by_ids(alg_id, implemented_alg_ids,
                                                    substitution_policy=substitution_policy,
                                                    id_dialect=IdDialect.integer)
>> 19
```

`get_preferred_substitute_algorithm_by_ids` will raise errors and warnings 
* Warnings: e.g., when LSODA substituted to CVODE
* Errors: e.g., when a tellurium doesn't implement a similar algorithm such as FBA, FVA, logical simulation
